src/lib.rs

```diff
@@ -140,15 +140,22 @@ impl Xor8 {
                     let h2 = filter.geth2(hash);
                     stack.push(keyindexvar);
 
-                    sets1[h1 as usize].xor_mask ^= hash;
-                    sets1[h1 as usize].count -= 1;
-                    if sets1[h1 as usize].count == 1 {
-                        q1.push(KeyIndex{index: h1, hash: sets1[h1 as usize].xor_mask})
+                    let mut s = unsafe {
```

Comment:
I feel this is early optimisation.

Reply:
It is an optimization, indeed (brings 3-5%% of performance improvements in my benchmarks).

I wouldn't agree about this being an _early_ one, though. Neither is this inherently less safe than what we have in `master` right now. Please let me explain.

1) I was carefully benchmarking and profiling the code on every step. There were some quite _idiomatic_ code improvements which I ended up rolling back, because they made the code perform worse, not better.
(For example, cleaning `sets.` with a loop seems less elegant when using `clear()` + `resize(desired_length, Default::default)`. Apparently, the loop still works better so I kept it the way it is now.)
In the code's current state the only actual heavy computation during construction happens in that `loop`, and in nested `while/for` loops inside there. Compiler is even smart enough to inline all the minor function calls (like `geth0`/`geth1`); it's all just a single loop on a hot code path. I'm actually reducing amount of array indirection (and involved checks) inside that nested loop, which has a measurable impact on the performance. I don't see any easy algorithmic/logical gains left in the code, so in my mind it's exactly the time to start squeezing relatively small gains out of our understanding of how compiler works.
So I would disagree, and say that this optimization happens exactly when it should be; it also doesn't make code harder to grasp in my view, which is a usual reason for objecting early optimisations, and a valid one. In this case, we just replace one method to access Vec by index with another, without any additional semantical complexity added.

2) Another reason to object this change would be introduction of `unsafe` into the codebase. But, while this is also something which makes sense in general, I don't think we're worsening situation here either.
`Index` is a funny trait; `[]` access might panic when out of bounds. So if we're not sure about our indices being in bounds for the Vec, we should use `Vec::get` and handle its `Option` return type, not just panicking and letting the code higher in the software stack deal with it.
But I don't think there's any risk of panicking on a bounds check here: `h[012]` and like are results of our `reduce` function, which strongly bounds the outputs to the allocated size of `sets.`; we don't resize `sets.` anywhere, so it's pretty safe to assume that our `h[012]` are always in its bounds.
This means that our `unsafe` code will never trigger any undefined behaviour, and we're all right here.
I probably should have added that as a comment to the `unsafe` block.

Will this change of mine be accepted if I rephrase my safety proof and add it as a comment?

Reply:
👍 I didn't notice the (2) line of reasoning while going through the change set. Thanks.